### PR TITLE
feat(video-player): add component (VIV-1486)

### PR DESCRIPTION
**Pull request by TaylorJ76**

### CODE REVIEW

File: libs/components/src/lib/video-player/video-player.scss [hunk 1]
```diff
@@ -0,0 +1,125 @@
+@use "../../../../shared/src/lib/sass/mixins/border-radius" as border-radius-variables;
+@import url("../../../../../node_modules/video.js/dist/video-js.css");
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:2]:
```suggestion
@import url("../../../../../node_modules/video.js/dist/video-js.css");
@use "../../../../../dist/libs/tokens/scss/tokens.constants" as constants;

```

File: libs/components/src/lib/video-player/video-player.scss [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+:host {
+	display: block;
+}
+
+.vjs-theme-vivid { 
+	position: relative;
+	border-radius: #{border-radius-variables.$border-radius-normal};
+	font: var(--vvd-typography-base);
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:11]:
```suggestion
	font: var(#{constants.$vvd-typography-base});
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	
+	.vjs-big-play-button {
+		width: 2.5em;
+		height: 2.5em;
+		margin-top: -1.25em;
+		margin-left: -1.25em;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:17]:
```suggestion
                 $big-button-size: 2.5em;
		inline-size: $big-button-size;
		block-size: $big-button-size;
		margin-block-start: -$big-button-size;
		margin-inline-start -$big-button-size;
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+	.vjs-poster, .vjs-tech {
+		overflow: hidden;
+		border-radius: #{border-radius-variables.$border-radius-normal};
+	}
+
+	.vjs-time-control { 
+		display: block; 
+		width: auto;
+		min-width: auto;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:30]:
```suggestion
		inline-size: auto;
		min-inline-size: auto; 
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+		overflow: hidden;
+		border-radius: #{border-radius-variables.$border-radius-normal};
+	}
+
+	.vjs-time-control { 
+		display: block; 
+		width: auto;
+		min-width: auto; 
+		padding: 0; 
+		line-height: 4em;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:32]:
```suggestion
		line-height: 4;
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+		padding: 0 0.5em;
+	}
+
+	.vjs-remaining-time { 
+		display: none; 
+	}
+
+	.vjs-control-bar {
+		width: calc(100% - 8px);
+		height: 4em;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:45]:
```suggestion
		inline-size: calc(100% - 8px);
		block-size: 4em;
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+			#{border-radius-variables.$border-radius-normal};
+	}
+
+	.vjs-menu {
+		left: -4em;
+		width: 12em;
+
+		ul {
+			border-radius:
+				#{border-radius-variables.$border-radius-normal}
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:72]: same as before :)

File: libs/components/src/lib/video-player/video-player.scss [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		border-radius: #{border-radius-variables.$border-radius-normal};
+
+		> .vjs-icon-placeholder::before {
+			line-height: 2.2;
+		}
+	}
+
+	.vjs-menu-button {
+		border-radius: 
+			#{border-radius-variables.$border-radius-none}
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:60]: not code related - but why are not we keeping the border-radius to the top as well - we don't have this kind of style in vivivd even when a menu is opened

File: libs/components/src/lib/video-player/video-player.scss [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+		border-radius: 
+			#{border-radius-variables.$border-radius-none} 
+			#{border-radius-variables.$border-radius-none} 
+			#{border-radius-variables.$border-radius-normal} 
+			#{border-radius-variables.$border-radius-normal};
+	}
+
+	.vjs-menu {
+		left: -4em;
+		width: 12em;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:68]:
```suggestion
		inset-inline-start: -4em;
		inline-size: 12em;
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+		li {
+			justify-content: start;
+			padding: 4px 8px;
+			text-align: left;;
+		}
+
+		.vjs-playback-rate & {
+			left: 0;
+			width: 4em;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:86]:
```suggestion
		inset-inline-start: 0;
		inline-size: 4em;
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+			li {
+				justify-content: center;
+				text-align: center;
+			}
+		}
+
+		li:not(.vjs-selected, .vjs-menu-title):hover {
+			background: #000;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:95]:
```suggestion
			background: var(#{constants.$vvd-color-canvas-text});
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+		}
+	}
+
+	.vjs-menu-button-popup .vjs-menu {
+		margin-bottom: 2.5em;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:100]:
```suggestion
		margin-block-end: 2.5em;
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	.vjs-playback-rate .vjs-playback-rate-value {
+		z-index: 1;
+		line-height: 2.5;
+	}
+
+	.vjs-button:hover {
+		background: #000;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:109]:
```suggestion
		background: var(#{constants.$vvd-color-canvas-text});
```

File: libs/components/src/lib/video-player/video-player.scss [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	.vjs-volume-bar {
+		margin: 1.9em 0.45em;
+	}
+
+	.vjs-volume-tooltip, .vjs-time-tooltip {
+		top: -5em;
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/video-player.scss:117]:
```suggestion
		inset-block-start: -5em;
```

File: libs/components/src/lib/video-player/README.md [hunk 15]
```diff
@@ -0,0 +1,222 @@
+# Video Player
+
+The video player component allows a user to play and control video content.
+
+```js
+<script type="module">import '@vonage/vivid/video-player';</script>
+```
+
+## Usage
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/README.md:9]:
```suggestion
## Members
```

File: libs/components/src/lib/video-player/README.md [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+
+### Video source
+
+#### src attribute
```
Comment (rachelbt) [inline libs/components/src/lib/video-player/README.md:13]:
```suggestion
#### src attribute

Type: string
Default: undefined
```
not sure this is the correct values - but this is missing for all the members

**Review by rachelbt — COMMENTED** [review]
nice :)
Went over the docs and the css.
The doc page is not aligned to other pages, I commented in some places but not all (laziness :) )

I think that the looks of the video player is not fully "vivid" especially the big play button.

the controls of the video in mobile are not responsive and causing horizontal scroll:
image

I left css comments and this is one in general - we are not using em in vivid, I see the need here to keep it responsive but it is a bit strange here, not sure what we need to do about it :)

**Review by YonatanKra — CHANGES_REQUESTED** [review]
One big component. Nice work!
A few changes to cope with Vivid standards.
image

File: libs/components/src/lib/enums.ts [hunk 17]
```diff
@@ -74,3 +74,10 @@ export enum AriaLive {
 	Polite = 'polite',
 	Assertive = 'assertive',
 }
+
+export enum SkipBy {
+	Zero = '0',
+	Five = '5',
+	Ten = '10',
+	Thirty = '30',
+}
```
[the remark this answers is not in the document]
Comment (TaylorJ76) [reply libs/components/src/lib/enums.ts:83]: The reason I didn't is that we have a request for audio player to have the same functionality.

File: libs/components/src/locales/en-GB.ts [hunk 18]
```diff
@@ -103,6 +104,103 @@ const enGB: Locale = {
 	},
 	splitButton: {
 		showMoreActionsLabel: 'Show more actions',
+	},
+	videoPlayer: {
+		'Audio Player': 'Audio Player',
```
Comment (RichardHelm) [inline libs/components/src/locales/en-GB.ts:109]: I think we can just import the locale from video.js to avoid copying it:
```suggestion
	videoPlayer: {
		videoJsLocale: <...imported from video.js...>,
```

File: libs/components/src/lib/video-player/README.md [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+### Poster image
+
+The poster image is displayed before the video has been interacted with. The `poster` attribute accepts a url string (absolute or relative) to an image.
+
+```html preview
+<style>
+	html {
+		max-width: 800px; /* for demo purposes */
+	}
+</style>
```
[the remark this answers is not in the document]
Comment (TaylorJ76) [reply libs/components/src/lib/video-player/README.md:54]:
```html preview 800px - Makes this preview panel 800px height, which doesn't look great.

**Comment by TaylorJ76** [comment]
> nice :) Went over the docs and the css. The doc page is not aligned to other pages, I commented in some places but not all (laziness :) )
> 
> I think that the looks of the video player is not fully "vivid" especially the big play button.
> 
> the controls of the video in mobile are not responsive and causing horizontal scroll: image
> 
> I left css comments and this is one in general - we are not using em in vivid, I see the need here to keep it responsive but it is a bit strange here, not sure what we need to do about it :)

I have sorted the responsiveness using container queries and converted `em`s to `px`s
And I have changed the big button to be smaller and hopefully more Vivid-like.

File: libs/components/src/lib/video-player/IconFontStyles.ts [hunk 20]
```diff
@@ -0,0 +1,7 @@
+/* eslint-disable max-len */
+export const iconFontStyles = `@font-face {
```
[the remark this answers is not in the document]
Comment (TaylorJ76) [reply libs/components/src/lib/video-player/IconFontStyles.ts:2]: I found a way to delete the icon font stuff all together. Now using the built in svgs

File: libs/components/src/lib/video-player/IconFontStyles.ts [hunk 21]
```diff
@@ -0,0 +1,7 @@
...
+    font-family: VideoJS;
```
[the remark this answers is not in the document]
Comment (TaylorJ76) [reply libs/components/src/lib/video-player/IconFontStyles.ts]: I have found a way to remove this completely.

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+			const bigPlayBtn = getBigPlayButton();
+			expect(bigPlayBtn?.classList.contains('vjs-hidden')).toBe(false);
+		});
+
+		it('should remove the redundant lang attribute', () => {
+			expect(element.shadowRoot?.querySelector('[lang]')).toBe(null);
+		});
+	});
+
+	describe('src', () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:65]: Are the below `it`s all the possible use cases for changing `src`? For instance 
- What happens when you remove `src`?
- What are valid values for `src`?
- How do you handle invalid `src`?

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 23]
```diff
@@ -0,0 +1,286 @@
...
+			await elementUpdated(element);
+			element.src = 'new-src.mp4';
+			await elementUpdated(element);
+			const bigPlayBtn = getBigPlayButton();
+			expect(bigPlayBtn?.classList.contains('vjs-hidden')).toBe(false);
+		});
+	});
+
+	describe('autoplay', () => {
+		it('should set the muted attribute on the video element', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:90]: This description doesn't seem to be the case anymore. Can you set a more accurate description please?

```suggestion
		it('should set reflect autoplay on the video element', async () => {
```

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 24]
```diff
@@ -0,0 +1,286 @@
...
+			)) as VideoPlayer;
+			await elementUpdated(element);
+			element.src = 'new-src.mp4';
+			await elementUpdated(element);
+			const bigPlayBtn = getBigPlayButton();
+			expect(bigPlayBtn?.classList.contains('vjs-hidden')).toBe(false);
+		});
+	});
+
+	describe('autoplay', () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:89]: Again, it seems like `autoplay` should have more tests. For instance 
- Does removal works the same?
- Can this be used as a programatic API? (e.g. element.autoplay = true;)

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 25]
```diff
@@ -0,0 +1,286 @@
...
+			const playbackRatesMenuItems = element.shadowRoot?.querySelectorAll('.vjs-playback-rate .vjs-menu li');
+			expect(playbackRatesMenuItems?.length).toBe(3);
+			if (playbackRatesMenuItems?.length === 3) {
+				expect(playbackRatesMenuItems[0].querySelector('.vjs-menu-item-text')!.textContent).toBe('1x');
+				expect(playbackRatesMenuItems[1].querySelector('.vjs-menu-item-text')!.textContent).toBe('0.5x');
+				expect(playbackRatesMenuItems[2].querySelector('.vjs-menu-item-text')!.textContent).toBe('0.25x');
+			}
+		});
+
+		it('should disable playback rates when passed an empty string', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:130]: According to the test, it should hide the playback rates, right?
```suggestion
		it('should hide playback rates when passed an empty string', async () => {
```

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 26]
```diff
@@ -0,0 +1,286 @@
...
+					<source src="${VIDEO_SRC}" type="video/mp4">
+				</${COMPONENT_TAG}>`
+			)) as VideoPlayer;
+			const playbackRate = element.shadowRoot?.querySelector('.vjs-playback-rate');
+			expect(playbackRate?.classList.contains('vjs-hidden')).toBe(true);
+		});
+	});
+
+	describe('captions', () => {
+		it('displays the captions button and menu', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:142]:
```suggestion
		it('should display the captions button and menu when caption tracks are set', async () => {
```

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 27]
```diff
@@ -0,0 +1,286 @@
...
+			element = (await fixture(
+				`<${COMPONENT_TAG} playback-rates="">
+					<source src="${VIDEO_SRC}" type="video/mp4">
+					<track 
+						kind="captions"
+						src="https://d2zihajmogu5jn.cloudfront.net/elephantsdream/captions.en.vtt" 
+						srclang="en" 
+						label="English" 
+						default>
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:151]: I suggest not to use external content. This can actually be replaced with any other string (even an empty one) because what matters is the number of tracks of `kind="captions"`.
```suggestion
						src=""
```

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 28]
```diff
@@ -0,0 +1,286 @@
...
+					<track 
+						kind="captions" 
+						src="https://d2zihajmogu5jn.cloudfront.net/elephantsdream/captions.en.vtt" 
+						srclang="fr" 
+						label="French">
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:156]:
```suggestion
						src=""
```

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+				</${COMPONENT_TAG}>`
+			)) as VideoPlayer;
+			await elementUpdated(element);
+			expect(element.shadowRoot!.querySelectorAll('[kind="captions"]').length).toBe(2);
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:160]: This expectation means nothing. All it is saying: "if I add two tracks of kind captions, there should be two elements of kind captions". Where's the captions button test? And probably the captions selection options test?

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 30]
```diff
@@ -0,0 +1,286 @@
...
+						default>
+    				<track 
+							kind="descriptions" 
+							src="https://d2zihajmogu5jn.cloudfront.net/elephantsdream/descriptions.en.vtt" 
+							label="French" 
+							srclang="fr">
+				</${COMPONENT_TAG}>`
+			)) as VideoPlayer;
+			await elementUpdated(element);
+			expect(element.shadowRoot!.querySelectorAll('[kind="descriptions"]').length).toBe(2);
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:183]: As above with captions - I cannot see how this expects a button. And as above, is the button all that's added? Or is there a menu to select from?

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 31]
```diff
@@ -0,0 +1,286 @@
...
+	describe('chapters', () => {
+		it('displays the chapters button and menu', async () => {
+			element = (await fixture(
+				`<${COMPONENT_TAG} playback-rates="">
+					<source src="${VIDEO_SRC}" type="video/mp4">
+					<track kind="chapters" src="https://d2zihajmogu5jn.cloudfront.net/elephantsdream/chapters.en.vtt">
+				</${COMPONENT_TAG}>`
+			)) as VideoPlayer;
+			await elementUpdated(element);
+			expect(element.shadowRoot!.querySelector('[kind="chapters"]')).not.toBe(null);
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:196]: As above - what should really happen. In this case, you are using `not`. This should be avoided most of the times, as negative assertions are a code smell

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 32]
```diff
@@ -0,0 +1,286 @@
...
+			element = (await fixture(
+				`<${COMPONENT_TAG} skip-by="30">
+					<source src="${VIDEO_SRC}" type="video/mp4">
+				</${COMPONENT_TAG}>`
+			)) as VideoPlayer;
+			const SkipBackwardBtn = element.shadowRoot?.querySelector('.vjs-skip-backward-30');
+			expect(SkipBackwardBtn?.classList.contains('vjs-hidden')).toBe(false);
+			const SkipForwardBtn = element.shadowRoot?.querySelector('.vjs-skip-forward-30');
+			expect(SkipForwardBtn?.classList.contains('vjs-hidden')).toBe(false);
+		});
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:211]: This test is more "correct" IMO:
```suggestion
		it('should modify skip button amount', async () => {
			element = (await fixture(
				`<${COMPONENT_TAG} skip-by="30">
					<source src="${VIDEO_SRC}" type="video/mp4">
				</${COMPONENT_TAG}>`
			)) as VideoPlayer;
			const [skipBackwardButtonWhenValueIs30, skipForwardButtonWhenValueIs30] = getSkipButtons(30);
			
			element.skipBy = 15;
			const [skipBackwardButtonWhenValueIs15, skipForwardButtonWhenValueIs15] = getSkipButtons(15);
			
			expect(skipBackwardButtonWhenValueIs30).toBeTruthy();
			expect(skipForwardButtonWhenValueIs30).toBeTruthy();
			expect(skipBackwardButtonWhenValueIs15).toBeTruthy();
			expect(skipForwardButtonWhenValueIs15).toBeTruthy();
			expect(getSkipButtons(30)[0]).toBeNull();
			expect(getSkipButtons(30)[1]).toBeNull();
		});
```
Notice I've used some helper function to make the code easier to read (it needs to be implemented - and of course reused in the rest of the tests using skip buttons :) ).

Due to its length, I'd even split it to 3 tests:
1. Initial state when no skip-by is given
2. Initial state when skip-by is given
3. Initial state when skip-by is given and value is changed dynamically (via JS API)
This will create 3 short tests that explain the API

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 33]
```diff
@@ -0,0 +1,286 @@
...
+
+		it('should disable skip by buttons when passed 0', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:213]: That's a great `edge case` test!

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 34]
```diff
@@ -0,0 +1,286 @@
...
+			element.addEventListener('ended', spy);
+			element.player.trigger('ended');
+			expect(spy).toHaveBeenCalledTimes(1);
+		});
+	});
+
+
+	describe('a11y', () => {
+		// skipped because there are aria tags (menus) inside video.js that are not correct
+		xit('should pass html a11y test', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:282]: This is a skipped test. Either remove it or make it pass :)

**Review by YonatanKra — CHANGES_REQUESTED** [review]
A few testing comments.

File: libs/components/src/lib/video-player/video-player.template.ts [hunk 35]
```diff
@@ -0,0 +1,27 @@
...
+ * The template for the VideoPlayer component.
+ *
+ * @param context - element definition context
+ * @public
+ */
+export const VideoPlayerTemplate: (
+	context: ElementDefinitionContext,
+	definition: FoundationElementDefinition
+) => ViewTemplate<VideoPlayer> = (
+	// context: ElementDefinitionContext
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.template.ts:22]: This can be removed

File: libs/components/src/lib/video-player/video-player.ts [hunk 36]
```diff
@@ -0,0 +1,179 @@
...
+				skipButtons: skipByValue > 0 ? skipButtons : false,
+				remainingTimeDisplay: { displayNegative: false },
+			},
+		};
+	}
+
+	/**
+	 * @internal
+	 */
+	initialiseVideo() {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.ts:155]: Two things:
1. This is funny, actually. Now I understand why `init` functions are `init` functions - to keep the peace between UK and US devs ;) My laptop screams at `initialise` and wants to turn it into `initialize` ;)
2. Can this be a private function as well? (`#initVideo`)

File: libs/components/src/lib/video-player/video-player.ts [hunk 37]
```diff
@@ -0,0 +1,179 @@
...
+		super.connectedCallback();
+		if (this.player) {
+			this.player.dispose();
+		}
+	}
+
+	/**
+	 * @internal
+	 */
+	getSettings() {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.ts:119]: Can this method be private (`#getSettings`)? 

Or even better, a private getter? 
```
#settings() {
...
}
```
Right now it is exposed and part of our API.

File: libs/components/tsconfig.json [hunk 38]
```diff
@@ -21,7 +21,8 @@
     "importHelpers": true,
     "allowSyntheticDefaultImports": true,
     "removeComments": true,
-    "composite": true
+    "composite": true,
+    "resolveJsonModule": true
```
Comment (YonatanKra) [inline libs/components/tsconfig.json:25]: Where do we import a `.json`?

File: libs/wrapper-gen/src/generator/polyfill/dom.ts [hunk 39]
```diff
@@ -23,4 +23,14 @@ global.matchMedia =
...
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+global.addEventListener = () => {};
+
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+global.removeEventListener = () => {};
+
+Object.defineProperty(global.document, 'readyState', {
+	value: 'complete',
+});
+
```
Comment (YonatanKra) [inline libs/wrapper-gen/src/generator/polyfill/dom.ts:35]: What does this solve?

File: libs/components/src/lib/video-player/video-player.ts [hunk 40]
```diff
@@ -0,0 +1,179 @@
...
+	 * @public
+	 * @remarks
+	 * HTML Attribute: skip-by
+	 */
+	@attr({attribute: 'skip-by', mode: 'fromView'}) skipBy: MediaSkipBy = MediaSkipBy.Ten;
+
+	/**
+	 * @internal
+	 */
+	player: any;
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.ts:96]: This can probably be a private property (`#player`)

File: libs/components/src/lib/video-player/video-player.ts [hunk 41]
```diff
@@ -0,0 +1,179 @@
...
+
+	/**
+	 * @internal
+	 */
+	videoEle: any;
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.ts:101]: Again, this should probably be private.

**Review by YonatanKra — CHANGES_REQUESTED** [review]
None tests related comments

File: libs/wrapper-gen/src/generator/polyfill/dom.ts (hunk 39 above)
Comment (TaylorJ76) [reply libs/wrapper-gen/src/generator/polyfill/dom.ts:35]: I think @RichardHelm added this as a fix to allow the docs site to deploy.

File: libs/components/tsconfig.json (hunk 38 above)
Comment (TaylorJ76) [reply libs/components/tsconfig.json:25]: In the locales files. We import the locales provided by video.js.

File: libs/components/src/lib/video-player/video-player.ts (hunk 40 above)
Comment (TaylorJ76) [reply libs/components/src/lib/video-player/video-player.ts:94]: We are using it in the tests to mock the events.

File: libs/components/src/lib/video-player/video-player.spec.ts (hunk 28 above)
Comment (TaylorJ76) [reply libs/components/src/lib/video-player/video-player.spec.ts:191]: Not sure why we need to remove the label

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 42]
```diff
@@ -0,0 +1,342 @@
...
+		});
+
+		it('should remove the lang attribute to avoid clash with vivid localization', () => {
+			expect(element.shadowRoot?.querySelector('[lang]')).toBe(null);
+		});
+	});
+
+	describe('src', () => {
+		describe('no src provided', () => {
+			it('should show the internal invalid src error message', async() => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:87]:
```suggestion
			it('should show the invalid src error when no src provided', async() => {
```

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 43]
```diff
@@ -0,0 +1,342 @@
...
+			expect(isBigPlayButtonVisible()).toBe(true);
+		});
+
+		it('should remove the lang attribute to avoid clash with vivid localization', () => {
+			expect(element.shadowRoot?.querySelector('[lang]')).toBe(null);
+		});
+	});
+
+	describe('src', () => {
+		describe('no src provided', () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:86]: can remove this describe as it holds only one usecase (see my suggestion for the `it`)

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 44]
```diff
@@ -0,0 +1,342 @@
...
+			it('should show the internal invalid src error message', async() => {
+				element = (await fixture(
+					`<${COMPONENT_TAG}></${COMPONENT_TAG}>`
+				)) as VideoPlayer;
+				await elementUpdated(element);
+				const noSrcErrorEl = element.shadowRoot!.getElementById('no-sources');
+				const dialogContentEl = getDialogContentEle();
+				expect(noSrcErrorEl?.classList.contains('vjs-hidden')).toBe(false);
+				expect(dialogContentEl!.textContent?.trim()).toBe('No compatible source was found for this media.');
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:95]: This is a bit confusing for me.
Creating the component with no `src` means two elements change? What's the connection between these two elements?

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 45]
```diff
@@ -0,0 +1,342 @@
...
+			});
+		});
+
+		it('should show the big play button by removing the vjs-hidden class', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:99]:
```suggestion
		it('should remove the vjs-hidden class when src is set', async () => {
```

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 46]
```diff
@@ -0,0 +1,342 @@
...
+		it('should be initialize with default state', async () => {
+			expect(element.src).toBe(undefined);
+			expect(element.poster).toBe(undefined);
+			expect(element.autoplay).toBe(false);
+			expect(element.loop).toBe(false);
+			expect(element.skipBy).toBe(MediaSkipBy.Ten);
+			expect(element.playbackRates).toBe(DEFAULT_PLAYBACK_RATES);
+		});
+
+		it('should show the big play button by removing the vjs-hidden class', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.spec.ts:76]: This is part of the `src` tests. You have the exact same test there.
This part is supposed to be the `clean init test` - so maybe test here that hidden class is set on first init with no params.

File: libs/components/src/lib/video-player/video-player.spec.ts (hunk 44 above)
Comment (TaylorJ76) [reply libs/components/src/lib/video-player/video-player.spec.ts:95]: When there is no src. Video.js appears as if it has loaded properly and displays the big play button.
I thought it would be more informative to the consumer if a message was displayed.

File: libs/components/src/lib/video-player/video-player.ts [hunk 47]
```diff
@@ -0,0 +1,213 @@
...
+			this.player = videojs(this.#videoEle, settings);
+			// removes lang="current" to avoid clash with vivid localization
+			this.shadowRoot!.querySelector('[lang]')!.removeAttribute('lang'); 
+			
+			this.player.on('play', () => this.$emit('play'));
+			this.player.on('pause', () => this.$emit('pause'));
+			this.player.on('ended', () => this.$emit('ended'));
+		} else {
+			if (noSourcesError) noSourcesError.classList.remove('vjs-hidden');
+		}
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/video-player.ts:208]: IMO most of it should go to the component's template. WDYT?

**Review by rachelbt — COMMENTED** [review]
Docs look great! 
So much more informative and tidy! 👍

**Review by rachelbt — COMMENTED** [review]
2 last ui test are nor rendering :(

File: libs/components/src/lib/video-player/ui.test.ts [hunk 48]
```diff
@@ -0,0 +1,159 @@
...
+	const promise = locator.evaluate((image: any) => image.complete || new Promise(f => image.onload = f));
+	await Promise.resolve(promise);
+
+	await page.waitForSelector('#video-player-controls .vjs-tech');
+	const videoEle = await page.$('#video-player-controls video');
+	const videoEle2 = await page.$('#video-player-controls-2 video');
+	const loadProgress = await page.$('#video-player-controls .vjs-load-progress');
+	const loadProgress2 = await page.$('#video-player-controls-2 .vjs-load-progress');
+	// hide elements that could lead to a flakey test
+	videoEle?.evaluate(element => element.style.visibility = 'hidden');
```
Comment (YonatanKra) [inline libs/components/src/lib/video-player/ui.test.ts:73]: Don't we want to test user behavior? You can emulate user clicks and actual interaction with the video...
Can we use the above instead of using `evaluate`?

File: libs/components/src/lib/video-player/locale.ts [hunk 49]
```diff
@@ -0,0 +1,97 @@
...
+	'{1} is loading.'?: string;
+	'No content'?: string ;
+	'Color'?: string;
+	'Opacity'?: string;
+	'Text Background'?: string;
+	'Caption Area Background'?: string;
+	'Skip forward {1} seconds'?: string;
+	'Skip backward {1} seconds'?: string;
+	'Playing in Picture-in-Picture'?: string;
+}
```
Comment (RichardHelm) [inline libs/components/src/lib/video-player/locale.ts:97]: We don't care about these so I think it should just be opaque:
```suggestion
export type VideoPlayerLocale = unknown;
```

File: libs/components/src/lib/video-player/video-player.ts [hunk 50]
```diff
@@ -0,0 +1,237 @@
...
+		const num = Number(numStr);
+		if (!isNaN(num)) ratesArray.push(num);
+	});
+	return ratesArray;
+}
+
+/**
+ * Base class for video-player
+ *
+ * @public
```
Comment (RichardHelm) [inline libs/components/src/lib/video-player/video-player.ts:23]:
```suggestion
 * @public
 * @component video-player
```

File: libs/components/src/lib/video-player/video-player.ts [hunk 51]
```diff
@@ -0,0 +1,237 @@
...
+	skipByChanged() {
+		if (this.player) {
+			this.#initVideo();
+		}
+	}
+
+	/**
+	 * @internal
+	 */
+	player: any;
```
Comment (RichardHelm) [inline libs/components/src/lib/video-player/video-player.ts:112]: We can prefix it at least:
```suggestion
	_player: any;
```

File: libs/components/src/lib/video-player/video-player.spec.ts [hunk 52]
```diff
@@ -0,0 +1,389 @@
...
+			const videoEle = getVideoEle();
+			expect(videoEle.loop).toBe(true);
+		});
+	});
+
+	function menuItemsMatchReverseOrderOfRates(playbackRates: string[], playbackRatesMenuItems: NodeListOf<Element>) {
+		return playbackRates.reverse().reduce((acc, rate, index) => {
+			return acc &&
+				playbackRatesMenuItems[index].querySelector('.vjs-menu-item-text')!.textContent === `${rate}x`;
+		}, true);
```
Comment (RichardHelm) [inline libs/components/src/lib/video-player/video-player.spec.ts:169]:
```suggestion
		return playbackRates.reverse().every((rate, index) =>
			playbackRatesMenuItems[index].querySelector('.vjs-menu-item-text')!.textContent === `${rate}x`;
		);
```

File: libs/components/src/lib/video-player/README.md [hunk 53]
```diff
@@ -0,0 +1,180 @@
...
+### Src
+
+A video source can be provided to the component using the `src` attribute.
+
+- Type: `string`
+- Default: `undefined`
+
+```html preview
+<style>
+	html { max-inline-size: 800px }
```
Comment (RichardHelm) [inline libs/components/src/lib/video-player/README.md:20]: How can consumers change the size of the video-player? Without changing <html>?
Also, the video player should not be display `inline`

**Review by RichardHelm — CHANGES_REQUESTED** [review]

File: libs/components/src/lib/video-player/README.md (hunk 53 above)
Comment (TaylorJ76) [reply libs/components/src/lib/video-player/README.md:20]: The Video is fluid and expands to fill the size of it's container, so consumers have control based on where it is placed in the page.

File: libs/components/src/lib/video-player/video-player.ts (hunk 51 above)
Comment (TaylorJ76) [reply libs/components/src/lib/video-player/video-player.ts:112]: We seem to have conflicting views on code conventions as Yonatan requested the prefix should be removed.

**Review by RichardHelm — APPROVED** [review]
